### ISSUE TITLE
Sampler 100% CPU >2h

### ISSUE THREAD

**Issue by andresriancho** [ecbftw/nanown#1]
I'm running my custom version of the sampler against a web app. It sends a lot of requests, and then:

```text
[1999/2000] Measured time: 33387000 - Sent data: -S-wU7ehYohG9yR-z65Wix
[1999/2000] Measured time: 32965000 - Sent data: -X-xV8ehYohG9yR-z65Wix
associate time: 2.4420907497406006
num_sent: 3, num_rcvd: 9
```

The first two lines are added by me for debugging requests. The last two are generated by `nanown`. The process has been running at 100% CPU usage for >2h, is that expected?

**Comment by andresriancho** [comment]
Problem seems to be inside `analyzeProbes`, adding some debug print:

```python
print('Analyzing probes...')
start = time.time()
num_probes = analyzeProbes(db)
```

And a new run output:
```text
Associating packets...
associate time: 0.7507262229919434
Analyzing probes...
num_sent: 3, num_rcvd: 9
```

**Comment by andresriancho** [comment]
PS: Running this from the smallest droplet inside digital ocean, maybe more CPU is required?

**Comment by andresriancho** [comment]
With a `sample_count` of 500 the process did finished and got https://github.com/ecbftw/nanown/issues/2

**Comment by andresriancho** [comment]
**Very** slow SQL query might be the reason for this issue:

```
def evaluateTrim(db, unusual_case, strim, rtrim):
    """
    septasummary and mad for each dist of differences
    """
    cursor = db.conn.cursor()
    query = """
      SELECT packet_rtt-(SELECT avg(packet_rtt) FROM probes,trim_analysis
                         WHERE sent_trimmed=:strim AND rcvd_trimmed=:rtrim AND trim_analysis.probe_id=probes.id AND probes.test_case!=:unusual_case AND sample=u.s AND probes.type in ('train','test'))
      FROM (SELECT probes.sample s,packet_rtt FROM probes,trim_analysis WHERE sent_trimmed=:strim AND rcvd_trimmed=:rtrim AND trim_analysis.probe_id=probes.id AND probes.test_case=:unusual_case AND probes.type in ('train','test')) u
    """
    # TODO: check for "N" in suspect field and return a flag
    print('Getting information from SQLite DB')
    
    params = {'strim': strim,
              'rtrim': rtrim,
              'unusual_case': unusual_case}
    cursor.execute(query, params)
    differences = [row[0] for row in cursor]
```

I'm seeing this in my command output:
```
(nanown)root@eulogias:/home/pablo/pch/nanown# ./analyze_packets do.db
Loading packets...
Done!
Processing packets...
Done!
Process packet output num_sent: 3, num_rcvd: 9
[1/27] Processing packets strim: 0 | rtrim 0
[2/27] Processing packets strim: 0 | rtrim 1
Processing packets...
Done!
[3/27] Processing packets strim: 0 | rtrim 2
Processing packets...
Done!
[4/27] Processing packets strim: 0 | rtrim 3
Processing packets...
Done!
[5/27] Processing packets strim: 0 | rtrim 4
Processing packets...
Done!
[6/27] Processing packets strim: 0 | rtrim 5
Processing packets...
Done!
[7/27] Processing packets strim: 0 | rtrim 6
Processing packets...
Done!
[8/27] Processing packets strim: 0 | rtrim 7
Processing packets...
Done!
[9/27] Processing packets strim: 0 | rtrim 8
Processing packets...
Done!
[10/27] Processing packets strim: 1 | rtrim 0
Processing packets...
Done!
[11/27] Processing packets strim: 1 | rtrim 1
Processing packets...
Done!
[12/27] Processing packets strim: 1 | rtrim 2
Processing packets...
Done!
[13/27] Processing packets strim: 1 | rtrim 3
Processing packets...
Done!
[14/27] Processing packets strim: 1 | rtrim 4
Processing packets...
Done!
[15/27] Processing packets strim: 1 | rtrim 5
Processing packets...
Done!
[16/27] Processing packets strim: 1 | rtrim 6
Processing packets...
Done!
[17/27] Processing packets strim: 1 | rtrim 7
Processing packets...
Done!
[18/27] Processing packets strim: 1 | rtrim 8
Processing packets...
Done!
[19/27] Processing packets strim: 2 | rtrim 0
Processing packets...
Done!
[20/27] Processing packets strim: 2 | rtrim 1
Processing packets...
Done!
[21/27] Processing packets strim: 2 | rtrim 2
Processing packets...
Done!
[22/27] Processing packets strim: 2 | rtrim 3
Processing packets...
Done!
[23/27] Processing packets strim: 2 | rtrim 4
Processing packets...
Done!
[24/27] Processing packets strim: 2 | rtrim 5
Processing packets...
Done!
[25/27] Processing packets strim: 2 | rtrim 6
Processing packets...
Done!
[26/27] Processing packets strim: 2 | rtrim 7
Processing packets...
Done!
[27/27] Processing packets strim: 2 | rtrim 8
Processing packets...
Done!
Evaluate trim 1/27
Getting information from SQLite DB
```

Using https://github.com/andresriancho/nanown/tree/develop

**Comment by ecbftw** [comment]
The `analyzeProbes` algorithm tries to determine which pair of sent/received packets should be used for `packet_rtt` measurements.  The algorithm efficiency can be improved, as can the heuristic.  Currently the algorithm runs `O(S*R)` where `S` is the typical number of data packets sent in the HTTP request and `R` is the typical number of data packets received. (The term "data packets" means the packets have non-0 payload length at the TCP layer.)  Your test seems to have quite a few sent and received packets (at least, more than I have been testing with) so there's a set of slow queries that are run 3*9=27 times.

There's a fairly straight-forward way to fix this by running some of the analyses in a short-circuit fashion, but I haven't had a chance to do it yet, since the algorithm details kept changing as I experimented with heuristics.  Notice the `analyzeProbes` function builds a dictionary called `evaluations` in a nested for loop.  Then it returns to analyze the contents of `evaluations` to determine which pair of packets to use.  If those loops were combined, we could perhaps convert this `O(S*R)` algorithm to a `O(max(S,R))` one, which would take your computation from 27 iterations to at most 9 iterations (though probably 4.5 iterations on average).

To speed things up in the short term, you should be able to skip this algorithm entirely by calling `analyzeProbes` with a specific `trim` parameter of `(0,0)`.  This is conservative and may not be optimal, but it will get you results quickly.

**Comment by ecbftw** [comment]
BTW- if you do happen to use reported times, then picking a trim parameter of (0,0) is completely reasonable, since this packet data won't be used anyway. =P

**Comment by ecbftw** [comment]
I have addressed this in 46cdd6efdb08512c4d129e387dd77812f5e19c21 by reorganizing analyzeProbes as I had described above.  I also sped up the routine that stores probe times temporarily in the database by about 4x by using sqlite3's executemany on the inserts, so you should see at least an order of magnitude faster analyzeProbes overall.  I'm sure more improvements can be made, but it should be adequate for now.
